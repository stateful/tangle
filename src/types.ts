import vscode from 'vscode';
import { Observable } from 'rxjs';

export interface Messenger {
    postMessage: (message: any) => void
}

export interface WebviewProvider {
    webview: Observable<vscode.Webview>;
    identifier: string;
}

/**
 * ToDo(Christian): eventually give this a more general type,
 * e.g. `Pick<EventTarget, 'dispatchEvent'>` and therefor stay
 * closer to native existing interfaces
 */
export interface Provider {
    onMessage: (listener: EventListener) => void;
    postMessage: (message: any) => Promise<void>;
}

export interface Receiver {
    /**
     * ToDo(Christian): eventually add `AddEventListenerOptions`
     */
    addEventListener(
        type: string,
        callback: (evt: Event) => void | null
    ): void;
}