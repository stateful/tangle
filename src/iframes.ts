import {
    of,
    map,
    take,
    toArray,
} from 'rxjs';

import type { Provider } from './types';
import BaseChannel from './channel';

export default class IFrameChannel<T> extends BaseChannel<HTMLIFrameElement, T> {
    public providers: HTMLIFrameElement[] = [];

    constructor (
        namespace: string,
        defaultValue?: Required<T>,
        /**
         * for testing purposes
         */
        private _window = window
    ) {
        super(namespace, defaultValue);
    }

    register (providers: HTMLIFrameElement[]) {
        this.providers.push(...providers);
        return of(...providers).pipe(
            take(providers.length),
            toArray(),
            map((ps: HTMLIFrameElement[]) => ps.map((p) => (<Provider>{
                onMessage: (listener) => {
                    this._window.onmessage = (ev) => listener(ev.data);
                },
                postMessage: (message) => {
                    if (!p.contentWindow) {
                        throw new Error('No content window found');
                    }
                    p.contentWindow.postMessage(message, '*');
                },
            }))),
            map(this._initiateBus.bind(this))
        );
    }

    attach () {
        return this._initiateClient(<Provider>{
            onMessage: (listener: EventListener) => {
                this._window.onmessage = (ev) => listener(ev.data);
            },
            postMessage: (message: any) => {
                if (!this._window.top) {
                    throw new Error('No top window found!');
                }
                this._window.top.postMessage(message, '*');
                return Promise.resolve();
            }
        });
    }
}
