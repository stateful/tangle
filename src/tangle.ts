import {
    BehaviorSubject,
    from,
    merge,
    Observable,
    of,
    Subject,
    bufferCount,
    filter,
    map,
    mergeMap,
    pluck,
    scan,
    share,
    withLatestFrom,
    throttleTime,
    first,
    EMPTY,
    firstValueFrom,
} from 'rxjs';
import type { Provider, Payload, Listener, RegisteredEvent, Context } from './types';

export class Client<T> {
    public readonly id: string = this._isBus ? 'bus' : Math.random().toString(36).substring(2);

    public readonly context: Observable<Context>;

    private readonly _outbound: Subject<Payload<T>>;
    private readonly _inbound: BehaviorSubject<Payload<T>>;
    private readonly _events: Subject<Payload<T>>;
    private readonly _notifer = new Subject<Context>();

    private readonly _transient: Observable<T>;
    private readonly _eventMap: Map<keyof T, RegisteredEvent<T>[]> = new Map();

    private readonly _context: Context = { clients: new Map() };

    constructor(
        public readonly namespace: string,
        public readonly providers: Provider[],
        public readonly defaultValue: T,
        private readonly _isBus: boolean = false
    ) {
        this._outbound = new Subject<Payload<T>>();
        this._inbound = new BehaviorSubject<Payload<T>>({ transient: this.defaultValue, context: this._context });
        this._events = new Subject<Payload<T>>();

        // bus keeps track of connections
        this._context.clients.set(this.id, this._isBus);
        this.context = this._registerContext();

        this._transient = this._register();
    }

    public dispose() {
        this._inbound.complete();
        this.removeAllListeners();
    }

    public get events() {
        return this._events.asObservable().pipe(share());
    }

    public get transient() {
        return this._transient;
    }

    public get state() {
        return this._inbound.value.transient;
    }

    /**
     * notify bus of new sandbox
     */
    public notify() {
        this._context.clients.set(this.id, true);
        this._notifer.next(this._context);
    }

    /**
     * collect and hold information about connected sandboxes
     */
    private _registerContext() : Observable<Context> {
        const context$ = this._inbound.pipe(
            pluck('context'),
            filter(Boolean),
            scan((acc, curr) => {
                if (Array.isArray(curr?.clients)) {
                    // deserialize array into es6 map
                    const _curr: Context = { clients: new Map(curr.clients) };
                    _curr.clients.forEach(((value, key) => acc.clients.set(key, value)));
                }
                return acc;
            }, this._context),
            share()
        );
        context$.subscribe(ctx => {
            this._context.clients = ctx.clients;
        });

        return context$;
    }

    /**
     * returns promise that fires when all sandboxes have connected
     */
    public whenReady(): Promise<Context>{
        return firstValueFrom(this.context.pipe(
            bufferCount(this.providers.length),
            map(() => this._context)
        ));
    }

    /**
     * share state with other sandboxes
     * @param state to share with other sandboxes
     */
    public broadcast(state: T) {
        this._outbound.next({ transient: state, context: this._context });
    }

    /**
     * event handler to act on changes to certain state properties
     * @param eventName state property
     * @param fn  handler to call once state of given property changes
     */
    public listen<K extends keyof T>(eventName: K, fn: Listener<T[K]>) {
        if (this._isBus) {
            fn(this.defaultValue[eventName]);
        }

        return this.events.pipe(
            pluck('transient'),
            pluck(eventName),
            filter((val) => val !== undefined)
        ).subscribe(fn);
    }

    /**
     * broadcast events with other sandboxes
     * @param eventName name of the event
     * @param payload   event payload
     */
    public emit<K extends keyof T>(eventName: K, payload: T[K]) {
        this._outbound.next({ event: { [eventName as string]: payload } } as any);
    }

    /**
     * listen to events shared within given namespace
     * @param eventName name of the event
     * @param fn        event handler
     */
    public on<K extends keyof T>(eventName: K, fn: Listener<T[K]>) {
        return this._registerEvent(eventName, fn);
    }

    /**
     * listen to a certain event shared within given namespace once
     * @param eventName name of the event
     * @param fn        event handler
     */
    public once<K extends keyof T>(eventName: K, fn: Listener<T[K]>) {
        return this._registerEvent(eventName, fn, true);
    }

    /**
     * listen to a certain event shared within given namespace once
     * @param subscription observable of event that should be unsubscribed
     */
    public off<K extends keyof T>(eventName: K, listener: Listener<T[K]>) {
        const events = this._eventMap.get(eventName) || ([] as RegisteredEvent<T>[]);
        events
            .filter(({ fn }) => fn === listener)
            .forEach(({ obs }) => obs.unsubscribe());
        return this;
    }

    /**
     * Get a listing the events for which the emitter has listeners.
     * The values in the array are strings or Symbols.
     * @returns `(string | symbol)[]`
     */
    public eventNames() {
        return [...this._eventMap.keys()];
    }

    /**
     * Returns the number of listeners listening to the event named eventName.
     * @param eventName The name of the event being listened for
     * @returns `integer`
     */
    public listenerCount<K extends keyof T>(eventName: K) {
        return this.listeners(eventName).length;
    }

    /**
     * Returns a copy of the array of listeners for the event named eventName.
     * @param eventName The name of the event being listened for
     * @returns `Function[]`
     */
    public listeners<K extends keyof T>(eventName: K) {
        return (this._eventMap.get(eventName) || ([] as RegisteredEvent<T>[]))
            .map(({ fn }) => fn);
    }

    /**
     * Removes all listeners, or those of the specified eventName.
     */
    public removeAllListeners() {
        const registeredEvents = [...this._eventMap.values()];
        for (const events of registeredEvents) {
            events.forEach(({ obs }) => obs.unsubscribe());
        }
        return this;
    }

    private _registerEvent<K extends keyof T>(eventName: K, fn: Listener<T[K]>, isOnce = false) {
        const events = this._eventMap.get(eventName) || ([] as RegisteredEvent<T>[]);
        const index = typeof eventName === 'string'
            ? eventName.toLocaleLowerCase()
            : eventName.toString().toLowerCase();

        const obs = this.events
            .pipe(
                pluck('event'),
                filter(Boolean),
                mergeMap((events) => from(Object.entries(events))),
                filter(([k]) => k.toLowerCase().indexOf(index) >= 0),
                map(([, v]) => v),
                (source) => {
                    if (!isOnce) {
                        return source;
                    }
                    return source.pipe(first());
                }
            )
            .subscribe(fn);

        // @ts-expect-error ToDo(Christian): fix TS error here
        events.push({ fn, obs });
        this._eventMap.set(eventName, events);
        return obs;
    }

    private _register(): Observable<T> {
        const providers = from(this.providers);
        const inGrouped$ = this._inbound;
        const outGrouped$ = this._outbound;

        const transientGrouped$ = merge(
            inGrouped$.pipe(pluck('transient')),
            outGrouped$.pipe(pluck('transient'))
        );

        const transient$ = merge(providers).pipe(
            this._fromProviders(),
            mergeMap(() => {
                return transientGrouped$.pipe(
                    scan(this._fold, this.defaultValue),
                    throttleTime(20),
                    mergeMap((transient) => {
                        const namespaced: Record<string, any> = {};
                        namespaced[this.namespace] = transient;
                        return of(transient);
                    })
                );
            }),
            bufferCount(this.providers.length),
            map((transients) => {
                return transients.reduce(this._fold, this.defaultValue);
            }),
            share()
        );

        if (!this._isBus) {
            inGrouped$.subscribe((state) => {
                if (state) {
                    this._events.next(state);
                }
            });
        }
        const event$ = this._isBus ? merge(outGrouped$, inGrouped$) : outGrouped$;
        event$
            .pipe(
                withLatestFrom(transient$),
                map(([event, transient]) => {
                    if (event.event) {
                        return { event: event.event };
                    }

                    if (event.transient) {
                        return { transient: { ...transient, ...event.transient } };
                    }
                }),
                map((combo) => {
                    if (combo === undefined) {
                        return EMPTY;
                    }

                    if (this._isBus) {
                        this._events.next({
                            event: combo.event,
                            transient: combo.transient || {} as T,
                            context: this._context
                        });
                    }

                    this.providers.forEach((provider) => {
                        provider.postMessage({ [this.namespace]: combo });
                    });
                })
            )
            .subscribe();

        if (!this._isBus) {
            this._notifer.pipe(map(_context => {
                // es6 map is not json serializable
                const context = { clients: Array.from(_context.clients.entries()) };
                return { context } as any; // special payload
            })).subscribe(payload => {
                this.providers.forEach((provider) => {
                    provider.postMessage({ [this.namespace]: payload });
                });
            });
        }

        transient$.subscribe();
        return transient$;
    }

    private _fromProviders() {
        return (source: Observable<Provider>) => {
            return source.pipe(
                map((provider) => {
                    provider.onMessage((payload) => {
                        const unpacked = (payload as any as Record<string, Payload<T>>)[this.namespace];
                        if (unpacked) {
                            this._inbound.next(unpacked);
                        }
                    });
                    return provider;
                })
            );
        };
    }

    private _fold(acc?: T, one?: T): T {
        if (!acc || !one) {
            return acc || one as T;
        }

        return { ...acc, ...one };
    }
}

export class Bus<T> extends Client<T> {
    constructor(namespace: string, providers: Provider[], defaultValue: T) {
        super(namespace, providers, defaultValue, true);
    }
}

export * from './types';
