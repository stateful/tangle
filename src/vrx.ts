import * as vscode from "vscode";
import {
  BehaviorSubject,
  from,
  merge,
  Observable,
  of,
  Subject,
  bufferCount,
  filter,
  map,
  mergeMap,
  pairwise,
  pluck,
  scan,
  share,
  startWith,
  take,
  toArray,
  withLatestFrom,
  distinctUntilKeyChanged,
  throttleTime,
} from "rxjs";

export interface WebviewProvider {
  webview: Observable<vscode.Webview>;
  identifier: string;
}

export interface Provider {
  onMessage: (listener: EventListener) => void;
  postMessage: (message: any) => Promise<void>;
}

export function wrapPanel(panel: vscode.WebviewPanel): WebviewProvider {
  return { webview: of(panel.webview), identifier: panel.viewType };
}

export function forWebviews<T>(
  namespace: string,
  defaultValue: T,
  wvProviders: WebviewProvider[]
): Observable<Bus<T>> {
  return merge(...wvProviders.map((wv) => wv.webview)).pipe(
    take(wvProviders.length),
    toArray(),
    map((wvs) => {
      return wvs.map((wv) => {
        const provider: Provider = {
          onMessage: (listener) => {
            wv.onDidReceiveMessage((event) => {
              listener(event);
            });
          },
          postMessage: (message) => {
            return new Promise((resolve) => {
              wv.postMessage(message).then(() => resolve());
            });
          },
        };
        return provider;
      });
    }),
    map((providers) => {
      return new Bus<T>(namespace, defaultValue, providers);
    })
  );
}

export function forDOM<T>(
  namespace: string,
  defaultValue: T,
  window: Window,
  vscode: { postMessage: (message: any) => void }
) {
  const provider: Provider = {
    onMessage: (listener: EventListener) => {
      window.addEventListener("message", (event) => listener(event.data));
    },
    postMessage: (message: any) => {
      vscode.postMessage(message);
      return Promise.resolve();
    },
  };

  return new Client<T>(namespace, defaultValue, [provider]);
}

export class Client<T> {
  protected readonly _outbound: Subject<T>;
  protected readonly _inbound: BehaviorSubject<T>;
  protected readonly _events: Subject<T>;

  protected readonly _transient: Observable<T>;

  constructor(
    public readonly namespace: string,
    public readonly defaultValue: T,
    public readonly providers: Provider[],
    private readonly _isBus: boolean = false
  ) {
    this._outbound = new Subject<T>();
    this._inbound = new BehaviorSubject<T>(this.defaultValue);
    this._events = new Subject<T>();

    this._transient = this.register();
  }

  public get events() {
    return this._events.asObservable().pipe(share());
  }

  public get transient() {
    return this._transient;
  }

  public broadcast(transient: T) {
    this._outbound.next(transient);
  }

  public listen(key: keyof T, fn: (...args: any[]) => void) {
    this.events
      .pipe(
        pluck(key),
        filter((val) => val !== undefined)
      )
      .subscribe(fn);
  }

  public on(eventName: string, fn: (...args: any[]) => void) {
    this.events
      .pipe(
        mergeMap((events) => from(Object.entries(events))),
        filter(([k]) => k.toLowerCase().indexOf(eventName.toLowerCase()) >= 0),
        map(([_, v]) => v)
      )
      .subscribe(fn);
  }

  public onAll(fn: (...args: any[]) => void) {
    this.events.subscribe(fn);
  }

  public subscribe(
    key: keyof T,
    fn: (...args: any[]) => void,
    distinctUntilChanged: boolean = false
  ) {
    if (distinctUntilChanged) {
      return this.transient
        .pipe(
          distinctUntilKeyChanged(key),
          pluck(key),
          filter((val) => val !== undefined)
        )
        .subscribe(fn);
    } else {
      return this.transient
        .pipe(
          pluck(key),
          filter((val) => val !== undefined)
        )
        .subscribe(fn);
    }
  }

  protected register(): Observable<T> {
    const providers = from(this.providers);
    const inGrouped$ = this._inbound.pipe(this.grouped());
    const outGrouped$ = this._outbound.pipe(this.grouped());

    const transientGrouped$ = merge(
      inGrouped$.pipe(pluck("transient")),
      outGrouped$.pipe(pluck("transient"))
    );

    const inEvent$ = inGrouped$.pipe(pluck("event"));
    inEvent$.subscribe((event) => {
      this._events.next(event);
    });

    const transient$ = merge(providers).pipe(
      this.fromProviders(),
      mergeMap((provider) => {
        return transientGrouped$.pipe(
          scan(this.fold, this.defaultValue),
          throttleTime(20),
          mergeMap((transient) => {
            const namespaced: any = {};
            namespaced[this.namespace] = transient;
            if (this._isBus) {
              return from(provider.postMessage(namespaced).then(() => transient));
            } else {
              return of(transient);
            }
          })
        );
      }),
      bufferCount(this.providers.length),
      map((transients) => {
        return transients.reduce(this.fold, this.defaultValue);
      }),
      share()
    );

    const outEvent$ = outGrouped$.pipe(pluck("event"));
    const event$ = this._isBus ? merge(outEvent$, inEvent$) : outEvent$;

    event$
      .pipe(
        withLatestFrom(transient$),
        map(([event, transient]) => {
          return { ...transient, ...event };
        }),
        map((combo) => {
          const namespaced: any = {};
          namespaced[this.namespace] = combo;

          this.providers.forEach((provider) => {
            provider.postMessage(namespaced);
          });
        })
      )
      .subscribe();

    transient$.subscribe();

    return transient$;
  }

  protected fromProviders() {
    return (source: Observable<Provider>) => {
      return source.pipe(
        map((provider) => {
          provider.onMessage((payload: any) => {
            const unpacked = payload[this.namespace];
            if (unpacked) {
              const payload = unpacked as T;
              this._inbound.next(payload);
            }
          });
          return provider;
        })
      );
    };
  }

  protected dedupe() {
    return (source: Observable<T>) => {
      return source.pipe(
        startWith(this.defaultValue),
        pairwise(),
        filter(([prev, curr]) => JSON.stringify(prev) !== JSON.stringify(curr)),
        map(([_, curr]) => curr)
      );
    };
  }

  protected fold(acc: T, one: T) {
    return { ...acc, ...one };
  }

  protected fromEntries(accum: T, [k, v]: [string, any]) {
    const r: any = { ...accum };
    r[k] = v;
    return r as T;
  }

  protected grouped() {
    return (source: Observable<T>) => {
      return source.pipe(
        map((payload) => {
          const entries = Object.entries(payload);
          const grouped: { transient?: T; event?: T }[] = entries.map((entry) => {
            const isPrefixed = entry[0].indexOf("on") === 0;
            const obj = this.fromEntries(this.defaultValue, entry);
            if (isPrefixed) {
              return { event: obj };
            }
            return { transient: obj };
          });
          return grouped;
        }),
        map((grouped) => {
          return {
            transient: grouped.reduce((accum, one) => {
              return { ...accum, ...one.transient };
            }, this.defaultValue),
            event: grouped.reduce((accum, one) => {
              return { ...accum, ...one.event };
            }, this.defaultValue),
          };
        })
      );
    };
  }
}

export class Bus<T> extends Client<T> {
  constructor(namespace: string, defaultValue: T, providers: Provider[]) {
    super(namespace, defaultValue, providers, true);
  }
}
