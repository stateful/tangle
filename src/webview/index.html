<html>
  <head>
    <title>VS Code RX</title>
    <base href="app-ext-path" />
  </head>
  <body>
    <script type="text/javascript" src="index.js"></script>
    <h3>VRX app-ext-identifier</h3>
    <button type="button" id="trigger" value="app-ext-identifier">Trigger</button>
    <h3></h3>
    <div id="root"></div>
  </body>
</html>
