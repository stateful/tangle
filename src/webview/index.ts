import { fromEvent, merge } from "rxjs";
import { map, pluck } from "rxjs/operators";
import * as Vrx from "vscoderx";

declare const acquireVsCodeApi: Function;

window.addEventListener("load", onload);

function onMsg(panel: string | null) {
  return (data: any) => {
    const div = document.createElement("div");
    div.appendChild(document.createTextNode(JSON.stringify(data)));
    const root = document.getElementById("root");
    if (root?.firstChild) {
      root?.insertBefore(div, root.firstChild);
      // root?.replaceChild(div, root.firstChild);
    } else {
      root?.appendChild(div);
    }
  };
}

function onload(event: Event) {
  const vscode = acquireVsCodeApi();

  const trigger = document.getElementById("trigger");
  if (trigger && vscode) {
    const label = trigger.getAttribute("value");
    const msg: any = { panel: label };
    if (typeof label === "string") {
      const eventName = `onPanel`;
      msg[eventName] = label;
    }

    // fromEvent(window, "message")
    //   .pipe(
    //     map((post: any) => post.data),
    //     pluck("vscoderx")
    //   )
    //   .subscribe(onMsg(label));

    const client = Vrx.forDOM("vscoderx", {}, window, vscode);
    // client.onAll(onMsg(label));
    client.on("panel", onMsg(label));
    client.on("countdown", onMsg(label));
    // client.transient.subscribe(onMsg(label));

    trigger.onclick = () => {
      client.broadcast(msg);
      // vscode.postMessage({ vscoderx: msg });
    };
  }
}
