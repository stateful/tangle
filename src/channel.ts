import { Observable } from 'rxjs';
import { Provider } from './types';
import { Client, Bus } from './tangle';

export default abstract class BaseChannel<U, T> {
    public abstract register (providers: U[]): Observable<Bus<T>>;

    constructor (
        private _namespace: string,
        private _defaultValue: Required<T>,
    ) {}

    public registerPromise (providers: U[]): Promise<Bus<T>> {
        return new Promise<Bus<T>>((resolve) => (
            this.register(providers).subscribe(resolve)
        ));
    }

    protected _initiateBus (providers: Provider[]) {
        return new Bus<T>(this._namespace, this._defaultValue, providers);
    }

    protected _initiateClient (provider: Provider) {
        return new Client<T>(this._namespace, this._defaultValue, [provider]);
    }
}
