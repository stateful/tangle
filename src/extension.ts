// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import { readFileSync } from "fs";
import path = require("path");
import * as vscode from "vscode";
import { map, of, Subject, take, timer } from "rxjs";
import * as Vrx from "./lib/vrx";

const webviewOptions = {
  enableScripts: true,
  retainContextWhenHidden: true,
};

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  const panel = vscode.window.createWebviewPanel(
    "column-one",
    "VRX column-one",
    vscode.ViewColumn.One,
    webviewOptions
  );
  const baseAppUri = panel.webview.asWebviewUri(
    vscode.Uri.file(path.join(context.extensionPath, "/dist/webview/"))
  );
  panel.webview.html = getHtml(context, baseAppUri.toString(), "column-one");

  const panelProviders: Vrx.WebviewProvider[] = [
    PanelViewProvider.register(context, "panel-one"),
    PanelViewProvider.register(context, "panel-two"),
    { webview: of(panel.webview), identifier: panel.viewType },
  ];

  // Subscribe to posts
  // const vrx = new Vrx.Bus<object>("vscoderx", {}, panelProviders);
  Vrx.forWebviews("vscoderx", {}, panelProviders).subscribe((bus) => {
    // Subscribe to events
    bus.on("panel", (msg) => console.log(`Listen to onPanel: ${msg}`));
    // vrx.onAll((msg) => console.log(`Listen to all: ${JSON.stringify(msg)}`));

    // Publish posts
    const countdown = 6;
    timer(2000, 8000)
      .pipe(
        take(countdown),
        map((i) => ({ onCountdown: countdown - 1 - i }))
      )
      .subscribe((msg) => {
        bus.broadcast(msg);
      });

    context.subscriptions.push(
      vscode.commands.registerCommand("vscoderx.emit", () => {
        bus.broadcast({ onCommand: "vscoderx.emit" });
      })
    );
  });
}

function getHtml(context: vscode.ExtensionContext, baseAppUri: string, identifier: string) {
  const re = /app-ext-identifier/g;
  const html = readFileSync(`${context.extensionPath}/src/webview/index.html`).toString("utf-8");
  return html.replace("app-ext-path", baseAppUri).replace(re, identifier);
}

export class PanelViewProvider implements vscode.WebviewViewProvider, Vrx.WebviewProvider {
  public view?: vscode.WebviewView;
  private _webview = new Subject<vscode.Webview>();

  constructor(
    private readonly _context: vscode.ExtensionContext,
    public readonly identifier: string
  ) {}

  resolveWebviewView(
    webviewView: vscode.WebviewView,
    webviewContext: vscode.WebviewViewResolveContext<unknown>,
    token: vscode.CancellationToken
  ): void {
    this.view = webviewView;
    const basePath = path.join(this._context.extensionPath, "/dist/webview/");

    webviewView.webview.options = {
      ...webviewOptions,
      localResourceRoots: [this._context.extensionUri],
    };
    const baseAppUri = webviewView.webview.asWebviewUri(vscode.Uri.file(basePath));
    webviewView.webview.html = getHtml(this._context, baseAppUri.toString(), this.identifier);
    this._webview.next(webviewView.webview);
  }

  public static register(context: vscode.ExtensionContext, identifier: string) {
    const panelProvider = new PanelViewProvider(context, identifier);
    context.subscriptions.push(
      vscode.window.registerWebviewViewProvider(identifier, panelProvider)
    );
    return panelProvider;
  }

  public get webview() {
    return this._webview.asObservable();
  }
}

// this method is called when your extension is deactivated
export function deactivate() {}
