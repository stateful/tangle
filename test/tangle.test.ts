import { test, expect } from 'vitest';
import { Observable } from 'rxjs';

import { Bus } from '../src/tangle';

interface Payload {
    foo: number
    bar: number
    [key: number | symbol]: number
}

test('has a transient getter', () => {
    const bus = new Bus<Payload>('testing', [], {} as Payload);
    expect(bus.transient instanceof Observable).toBe(true);
});

test('allows to get state value', () => {
    const bus = new Bus<Payload>('testing', [], { foo: 0, bar: 0 } as Payload);
    expect(bus.state).toMatchSnapshot();
});

test('has a list if event names', () => {
    const noop = () => { /** */ };
    const bus = new Bus<Payload>('testing', [], {} as Payload);
    bus.on('foo', noop);
    bus.on('bar', noop);

    const sym = Symbol('symbol');
    bus.on(sym, noop);
    expect(bus.eventNames()).toMatchSnapshot();
});

test('allows to get listener count', () => {
    const noop = () => { /** */ };
    const bus = new Bus<Payload>('testing', [], {} as Payload);
    bus.on('foo', noop);
    bus.on('foo', noop);
    bus.on('foo', noop);
    bus.on('bar', noop);

    const sym = Symbol('symbol');
    bus.on(sym, noop);
    bus.on(sym, noop);
    expect(bus.listenerCount('foo')).toBe(3);
    expect(bus.listenerCount('bar')).toBe(1);
    expect(bus.listenerCount(sym)).toBe(2);
    // @ts-expect-error wrong key
    expect(bus.listenerCount('foobar')).toBe(0);
});

test('allows to get listeners of certain event', () => {
    const noop = () => { /** */ };
    const bus = new Bus<Payload>('testing', [], {} as Payload);
    bus.on('foo', noop);
    const fn = bus.listeners('foo')[0];

    expect(fn).toEqual(noop);
});
