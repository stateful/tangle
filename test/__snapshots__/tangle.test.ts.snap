// Vitest Snapshot v1

exports[`allows to get state value 1`] = `
{
  "bar": 0,
  "foo": 0,
}
`;

exports[`has a list if event names 1`] = `
[
  "foo",
  "bar",
  Symbol(symbol),
]
`;
